### PR TITLE
Feature/mapping shape

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
rename exclusive shape to inverse shape and add extrude shape for generate wall boundary.

### CODE REVIEW

File: tests/2d_examples/test_2d_particle_generator_single_resolution/particle_generator_single_resolution.cpp
[the remark this answers is not in the document]
Comment (yyc84) [reply tests/2d_examples/test_2d_particle_generator_single_resolution/particle_generator_single_resolution.cpp:1]: image
When I only use original geometry to generate levelset and particles, it seems pretty ok,

**Comment by Xiangyu-Hu** [comment]
now the level set seems ok.
Screenshot from 2023-07-07 22-49-45

File: src/shared/geometries/mapping_shape.h [hunk 1]
```diff
@@ -35,29 +35,69 @@
...
     virtual bool checkContain(const Vecd &probe_point, bool BOUNDARY_INCLUDED = true) override
     {
         return !BaseShapeType::checkContain(probe_point);
     };
 };
+
+/**
+ * @class ExtrudeShape
+ * @brief A template shape which define the region by expanding the geometry surface with given thickness.
+ * @brief Positive thickness will extend the shape and negative thickness will shrink the shape.
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/geometries/mapping_shape.h:65]: no. It extend a volume model.

File: src/shared/geometries/mapping_shape.h [hunk 2]
```diff
@@ -35,29 +35,69 @@
...
+ */
+template <class BaseShapeType>
+class ExtrudeShape : public BaseShapeType
+{
+    Real thickness_;
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/geometries/mapping_shape.h]: It gives the thickness.  It is a real number.

File: tests/2d_examples/test_2d_particle_generator_single_resolution/particle_generator_single_resolution.cpp
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply tests/2d_examples/test_2d_particle_generator_single_resolution/particle_generator_single_resolution.cpp:1]: It is correct now.

**Comment by Xiangyu-Hu** [comment]
@yyc84 if it is ok. You can approve the pull request.

**Comment by yyc84** [comment]
I have checked it, and this time the level-set seems quite smooth.

**Comment by Nadun19** [comment]
I tried to modify particle_relaxation_single_resolution_3D.cpp:

`class SolidBodyFromMesh : public ComplexShape
{
  public:
    explicit SolidBodyFromMesh(const std::string &shape_name) : ComplexShape(shape_name)
    {
        TriangleMeshShapeSTL original_stl(full_path_to_file, translation, scaling);
        add<ExtrudeShape<TriangleMeshShape>>(4.0 * dp_0, original_stl);
    }
};`

But it does not compile, I am not entirely sure how to use the TriangleMeshShape class properly.

**Comment by Xiangyu-Hu** [comment]
3d case
WM-Screenshots-20230710120216

**Pull request merged**